feat: use-template for forms

wanlingt opened the pull request:
## Problem
<!-- What problem are you trying to solve? What issue does this close? -->

Closes #5161
## Solution
<!-- How did you solve the problem? -->

**Breaking Changes** 
<!-- Does this PR contain any backward incompatible changes? If so, what are they and should there be special considerations for release? -->
- [x] No - this PR is backwards compatible  

**Features**:
- Public forms can be used as a template

**Details**:
- Added the use-template link to `ShareFormModal`. This input field is disabled if the form is closed.
- Added a new component `TemplateFormPage`, which is similar in structure to `PreviewFormPage`. This also uses a new `TemplateFormProvider`, which only differs from `PreviewFormProvider` in the query used (`useFormTemplate`)
- Use `DuplicateFormModal` to duplicate the template form. The `formId` is now injected into `DupeFormWizardProvider` at a higher level, so that `DuplicateFormModal` can be used across workspace rows and the template form page.
- Use `z-index` in form page's `MiniHeader` instead of `Portal` to enable stacking of the sticky headers

**Improvements**:

- Updated the design of `PreviewFormPage`

## Before & After Screenshots

**AFTER**:
<!-- [insert screenshot here] -->

https://user-images.githubusercontent.com/56983748/202135080-868110c0-1804-45b3-847c-894eafc6803c.mov


## Tests
<!-- What tests should be run to confirm functionality? -->
New stories have been added.

Use-template:
- [x] As a logged in admin, visit the use-template endpoint of a public form `admin/form/:formId/use-template`. Click on the button 'Use this template'. The form title should be prefixed with '[Template]' and you should be able to duplicate the form successfully.
- [x] As a logged in admin, visit the use-template endpoint of a **public** form that you're not the owner of. You should be able to create a duplicate of the form template successfully.
- [x] As a logged in admin, visit the use-template endpoint of a **private** form that you're not the owner of. You should see the admin forbidden error page.
- [x] As a public user, visit the use-template endpoint of a public form. You should be redirected to the login page.

wanlingt committed:
feat: add use-template link to share form modal

wanlingt committed:
feat: update preview mode design

wanlingt committed:
feat: add use-template v3 route

wanlingt committed:
feat: add template form page

wanlingt committed:
feat: use DuplicateFormModal for form template creation

wanlingt committed:
fix: remove unnecessary string

wanlingt committed:
feat: sticky template preview banner

wanlingt committed:
fix: change template preview heading

wanlingt committed:
feat: add mobile drawer

wanlingt committed:
feat: add margin to mini header in template preview mode

wanlingt committed:
feat: fix position of mini header by replacing portal with z-index

wanlingt committed:
test: add stories

wanlingt committed:
fix: remove unused Portal import

wanlingt committed:
feat: add useFormTemplate query

wanlingt committed:
feat: add TemplateFormProvider

wanlingt committed:
feat: add admin forbidden error page

wanlingt committed:
fix: remove console.log

wanlingt committed:
ref: update comments

justynoh commented on frontend/src/features/admin-form/common/AdminViewFormService.ts:
```diff
@@ -69,6 +69,35 @@ export const previewForm = async (
+  formId: string,
+): Promise<PreviewFormViewDto> => {
+  return ApiService.get<PreviewFormViewDto>(
+    `${ADMIN_FORM_ENDPOINT}/${formId}/use-template`,
```
Should we reference the `PUBLICFORM_USETEMPLATE_ROUTE` constant in `../../../constants/routes` instead?

wanlingt replied:
[d147cf9](https://github.com/opengovsg/FormSG/pull/5377/commits/d147cf94f10e80bfad26efb58c95c08305369a6a)

justynoh commented on frontend/src/features/admin-form/share/ShareFormModal.tsx:
```diff
@@ -161,6 +161,32 @@ export const ShareFormModal = ({
               </Skeleton>
             </FormControl>
+            <FormControl isReadOnly>
+              <FormLabel isRequired useMarkdownForDescription>
```
supernit: since we don't have a description, we probably don't need `useMarkdownForDescription` here haha

separately, actually, i wonder if we _should_ have a description that reminds admins that if their form is closed, other admins will not be able to use-template

wanlingt replied:
yeah good idea! i've removed `useMarkdownForDescription` for now in this commit [d147cf9](https://github.com/opengovsg/FormSG/pull/5377/commits/d147cf94f10e80bfad26efb58c95c08305369a6a)

@staceytan1998 shall we add the reminder description for admins? maybe something like 'If your form is closed, other admins will not be able to view a template of your form"

wanlingt replied:
added the reminder to the share form modal [ec6c92e](https://github.com/opengovsg/FormSG/pull/5377/commits/ec6c92e187fa1c976db8d50eb36868208e6c550b)

justynoh commented on frontend/src/features/admin-form/share/ShareFormModal.tsx:
```diff
@@ -161,6 +161,32 @@ export const ShareFormModal = ({
+                      <CopyButton
+                        colorScheme="secondary"
+                        stringToCopy={`${shareLink}/use-template`}
+                        aria-label="Copy use-template form link"
```
nit: For a11y, should we rephrase "use-template" in `aria-label`?

wanlingt replied:
[d147cf9](https://github.com/opengovsg/FormSG/pull/5377/commits/d147cf94f10e80bfad26efb58c95c08305369a6a)

justynoh commented on frontend/src/features/admin-form/share/ShareFormModal.tsx:
```diff
@@ -161,6 +161,32 @@ export const ShareFormModal = ({
+                    <InputRightElement>
+                      <CopyButton
+                        colorScheme="secondary"
+                        stringToCopy={`${shareLink}/use-template`}
```
nit: should we factor out the link, and also reference the route constant again?

wanlingt replied:
yes good catch! [d147cf9](https://github.com/opengovsg/FormSG/pull/5377/commits/d147cf94f10e80bfad26efb58c95c08305369a6a)

justynoh commented on frontend/src/features/admin-form/common/components/PreviewFormBanner/PreviewFormBanner.tsx:
```diff
@@ -1,39 +1,183 @@
-            <Icon aria-hidden as={BxsEditAlt} fontSize="1rem" mr="0.75rem" />
-            <Text textStyle="subhead-2">Edit form</Text>
+            <Icon aria-hidden as={BiShow} fontSize="1rem" mr="0.75rem" />
+            <Text textStyle="subhead-3">Template Preview</Text>
```
Based on figma, templates should say "Template preview" but form preview should say "Form preview", hahah

wanlingt replied:
[d147cf9](https://github.com/opengovsg/FormSG/pull/5377/commits/d147cf94f10e80bfad26efb58c95c08305369a6a)

justynoh commented on frontend/src/features/workspace/components/DuplicateFormModal/DupeFormWizardProvider.tsx:
```diff
@@ -53,7 +54,12 @@ export const useDupeFormWizardContext = (): CreateFormWizardContextReturn => {
-      title: makeDuplicateFormTitle(previewFormData.form.title, dashboardForms),
+      title: makeDuplicateFormTitle(
+        isTemplate
+          ? `Template_${previewFormData.form.title}`
```
not about your PR, but omggg why are we still prefixing with "Template_" so gross :(

wanlingt replied:
@staceytan1998 any comments HAHA

wanlingt replied:
we'll change this to [Template] after we allow special characters in the form title

wanlingt replied:
[3724a12](https://github.com/opengovsg/FormSG/pull/5377/commits/3724a12d3e816e7ee2988d401dfb0cf0915f1d72) updated yay

justynoh commented on frontend/src/features/workspace/components/DuplicateFormModal/DupeFormWizardProvider.tsx:
```diff
@@ -14,17 +14,18 @@ import {
+      formId,
       // Stop querying once submissionData is present.
-      /* enabled= */ !!activeFormMeta,
+      /* enabled= */ !!formId,
```
because of this check, i wonder if actually we actually need `formId` to be provided. could we possibly make `formId` an optional argument, then in `WorkspaceRowsContext.tsx` we don't have to handle the undefined case?

```javascript
        formId={activeFormMeta?._id}

wanlingt replied:
i left it as is as it's technically needed for in the queries `usePreviewForm(formId)` / `useTemplateForm(formId)`

justynoh replied:
haha as in, we can leave it undefined, then in the dupe form wizard provider do 

(but either way is ok la. is up to preference haha)

```javascript
const useDupeFormWizardContext = (formId = '', isTemplate) => ...

justynoh commented on frontend/src/features/workspace/components/DuplicateFormModal/DupeFormWizardProvider.tsx:
```diff
@@ -14,17 +14,18 @@ import {
-  const { activeFormMeta } = useWorkspaceRowsContext()
 
   const { data: previewFormData, isLoading: isPreviewFormLoading } =
     usePreviewForm(
```
are we sure that this works as intended if we only call `usePreviewForm` for the template feature? In the backend, `handlePreviewAdminForm` checks for the current user, while `handleGetTemplateForm` does not. So when a different admin tries to dupe the form, this might not work.

wanlingt replied:
i cri u are right

wanlingt replied:
made the queries somewhat conditional here [db5601b](https://github.com/opengovsg/FormSG/pull/5377/commits/db5601b5a604971618758486ec619167b987d3b7), but not sure if it's the best way to do so, lmk what you think!

wanlingt committed:
ref: naming fixes

wanlingt committed:
feat: add useFormTemplate query to DupeFormWizardProvider

wanlingt committed:
style: copy and formatting changes

wanlingt committed:
fix: use FormNotFound error page

wanlingt committed:
test: update stories with use-template path

justynoh commented on frontend/src/features/workspace/components/DuplicateFormModal/DupeFormWizardProvider.tsx:
```diff
@@ -24,13 +27,23 @@ export const useDupeFormWizardContext = (
   const { data: previewFormData, isLoading: isPreviewFormLoading } =
     usePreviewForm(
       formId,
-      // Stop querying once submissionData is present.
-      /* enabled= */ !!formId,
+      // Stop querying if formId does not exist or if it's in view template mode
+      /* enabled= */ !!formId || !isTemplate,
     )
 
+  const { data: templateFormData, isLoading: isTemplateFormLoading } =
+    useFormTemplate(
+      formId,
+      // Stop querying if formId does not exist or if it's not in preview mode
+      /* enabled= */ !!formId || isTemplate,
+    )
+
+  const data = templateFormData || previewFormData
+  const isLoading = isTemplateFormLoading || isPreviewFormLoading
```
Suggestion:

```javascript

const { data, isLoading } = isTemplate ? useFormTemplate(formId, !!formId) : usePreviewForm(formId, !!formId)

wanlingt replied:
cant do that :( `React Hook "useFormTemplate" is called conditionally. React Hooks must be called in the exact same order in every component render`

justynoh replied:
oh ya LOL lame, sure then hahaha that's fine

justynoh commented on frontend/src/features/admin-form/template/TemplateFormProvider.tsx:
```diff
@@ -65,7 +65,8 @@ export const TemplateFormProvider = ({
   const isFormNotFound = useMemo(() => {
     return (
-      error instanceof HttpError && (error.code === 404 || error.code === 410)
+      error instanceof HttpError &&
```
not about this line, but there's a linter complaint in this file:

```
Check warning on line 21 in frontend/src/features/admin-form/template/TemplateFormProvider.tsx

GitHub Actions / frontend_lint
'AdminForbiddenErrorPage' is defined but never used
```

justynoh reviewed (approved):
lgtm

justynoh commented on frontend/src/features/admin-form/template/TemplateFormProvider.tsx:
```diff
@@ -0,0 +1,188 @@
+import { HttpError } from '~services/ApiService'
+import { FormFieldValues } from '~templates/Field'
+
+import AdminForbiddenErrorPage from '~pages/AdminForbiddenError'
```
there's a linter complaint here:

```
Check warning on line 21 in frontend/src/features/admin-form/template/TemplateFormProvider.tsx

GitHub Actions / frontend_lint
'AdminForbiddenErrorPage' is defined but never used
```

wanlingt replied:
ooh yes tq [bccfb1c](https://github.com/opengovsg/FormSG/pull/5377/commits/bccfb1c8c07b6fa6d7a8500c5beed4760959c675)

wanlingt committed:
fix: remove unused import

wanlingt committed:
feat: map angular route

wanlingt committed:
fix: use admin form routing

wanlingt committed:
feat: disallow non-valid-utf-8 encoded characters in form title

wanlingt committed:
fix: update template prefix to be [Template]

wanlingt commented:
@justynoh rerequesting review for the routing and new form title validation portions!

timotheeg commented on frontend/src/features/admin-form/common/components/PreviewFormBanner/PreviewFormBanner.tsx:
```diff
@@ -1,39 +1,197 @@
+      {isTemplate ? <StickyPreviewHeader isOpen={isOpen} /> : null}
+      <PreviewFormBanner isTemplate={isTemplate} />
+      {
+        /* Sentinel to know when sticky navbar is starting */
+
+        <Waypoint topOffset="64px" onPositionChange={handlePositionChange} />
```
I see we already have a Sentinel waypoint in the mini header. Do that and this preview banner ever show together? If yes do the waypoints work well together?

wanlingt replied:
yup, the sticky use-template preview banner and the mini header show up in use-template mode. they seem to work well together! the waypoints are specific to each element

timotheeg commented on src/app/models/form.server.model.ts:
```diff
@@ -163,10 +167,13 @@ const compileFormModel = (db: Mongoose): IFormModel => {
-        ],
+        validate: {
+          // validation fails if regex pattern exists in the title
+          validator: function (v: string) {
+            return !UNICODE_ESCAPED_REGEX.test(v)
+          },
```
/super nit: use shorter syntax for a simple function like this
```suggestion
          validator: v => !UNICODE_ESCAPED_REGEX.test(v),
```

timotheeg replied:
More importantly: are you sure this test is correct here? `UNICODE_ESCAPED_REGEX` detects the incorrect utf8 characters by inspecting the json-encoded version of the value, but `v` here is the raw string value, no?

timotheeg commented on src/app/modules/form/admin-form/admin-form.controller.ts:
```diff
@@ -136,7 +136,24 @@ const createFormValidator = celebrate({
-      .unknown(true),
+      .unknown(true)
+      .custom((value, helpers) => {
+        // If there are unicode-escaped characters are not valid utf-8 encoded,
```
/nit grammar in this comment reads funny?

justynoh commented:
@wanlingt can I suggest that we merge the last two commits (allowing special characters in form titles and updating template prefix) in a separate PR instead? So that we get better separation of features if we have to rollback either one

timotheeg commented on src/app/models/form.server.model.ts:
```diff
@@ -163,10 +167,13 @@ const compileFormModel = (db: Mongoose): IFormModel => {
       title: {
         type: String,
-        validate: [
-          /^[a-zA-Z0-9_\-./() &`;'"]*$/,
```
/justMusing: Ooohh, so we did allow some special characters even previously, we could have set the prefix as `(Template) `  and reduce the changeset 🤔

timotheeg commented on src/app/modules/form/admin-form/admin-form.middlewares.ts:
```diff
@@ -26,5 +27,23 @@ export const updateSettingsValidator = celebrate({
+      if (UNICODE_ESCAPED_REGEX.test(valueStr)) {
+        return helpers.message({
+          custom:
+            'Please check that there are no improperly encoded characters in your input',
```
I wonder if folks would understand this error message 🤔 . Since we **know** there are invalid UTF8 characters in the input, we should probably say it outright.
```suggestion
            'There are invalid characters in your input',
```

timotheeg commented:
> @wanlingt can I suggest that we merge the last two commits (allowing special characters in form titles and updating template prefix) in a separate PR instead? So that we get better separation of features if we have to rollback either one

Good idea! This PR could do a simpler change of allowing 2 new characters in the title `[]` (by expanding the regex just a bit), and a secondary PR to allow unrestricted unicode content could follow. This whole process might take a bit more than  the 1 point we planned though (or was it zero even? 😅 )

wanlingt committed:
Revert "fix: update template prefix to be [Template]"

This reverts commit 3724a12d3e816e7ee2988d401dfb0cf0915f1d72.

wanlingt committed:
Revert "feat: disallow non-valid-utf-8 encoded characters in form title"

This reverts commit 6ca53ae74dcd14a9371a601958ba0539e0440fa9.

wanlingt committed:
feat: allow square brackets in form title

wanlingt commented:
> @wanlingt can I suggest that we merge the last two commits (allowing special characters in form titles and updating template prefix) in a separate PR instead? So that we get better separation of features if we have to rollback either one

yes! good suggestion. I've reverted the last two commits in this PR, will open a separate PR for them :) so now the only change i've added is to allow square brackets in the title [f7f7c2c](https://github.com/opengovsg/FormSG/pull/5377/commits/f7f7c2caa24dd1f4fa7d914891c5e185831f8aec)

@timotheeg fyi too!

karrui commented on frontend/src/features/admin-form/template/TemplateFormProvider.tsx:
```diff
@@ -0,0 +1,187 @@
+    [data?.form, data?.spcpSession],
+  )
+
+  const { submitEmailModeFormMutation, submitStorageModeFormMutation } =
+    usePreviewFormMutations(formId)
```
We should not allow submission on templates - using the preview endpoint may fail if the form does not belong to the user. Just disable altogether imho

wanlingt replied:
good point! opened a new pr [here](https://github.com/opengovsg/FormSG/pull/5443) to disable submission
